File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int dp[1001];
    int n;
    
    int helper(int curr,vector<int>&arr){
        
        if(curr>=n)
            return 0;
        
        if(dp[curr]!=-1)
            return dp[curr];
        
        return dp[curr] = arr[curr] + min(helper(curr+1,arr) , helper(curr+2,arr));
        
    }
    int minCostClimbingStairs(vector<int>& cost) {
        memset(dp,-1,sizeof dp);
        n  = cost.size();
        
        return min(helper(0,cost), helper(1,cost));
    }
};

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.java
class Solution {
    private int n;
    private int[]dp;
    
    private int helper(int curr,int[] cost){
        
        if(curr>=n)
            return 0;
        
        if(dp[curr]!=-1)
            return dp[curr];
        
        
        return dp[curr] = cost[curr] + Math.min(helper(curr+1,cost), helper(curr+2,cost));
        
    }
    public int minCostClimbingStairs(int[] cost) {
        
        n = cost.length;
        dp = new int[n+1];
        Arrays.fill(dp,-1);
        helper(0,cost);
        return Math.min(dp[0],dp[1]);
        
        
    }
}